aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,397 @@
+#pragma once
+
+#include <ATen/core/stack.h>
+#include <c10/util/TypeList.h>
+#include <ATen/core/boxing/kernel_functor.h>
+#include <ATen/core/boxing/kernel_function.h>
+#include <ATen/core/boxing/kernel_lambda.h>
+
+namespace c10 {
+
+namespace detail {
+template<class Return, class... Args> struct boxAndCallBoxedFunc;
+}
+
+/**
+ * KernelFunction is similar to std::function but stores a kernel function.
+ * You can create a KernelFunction from a boxed or unboxed function/functor/lambda
+ * and call it in a boxed or unboxed way. If the way it was created doesn't
+ * match the way it was called, it will do boxing or unboxing as necessary.
+ */
+class CAFFE2_API KernelFunction final {
+public:
+  using BoxedKernelFunction = void(OperatorKernel*, Stack*);
+
+  KernelFunction()
+  : functorFactory_()
+  , functor_(nullptr)
+  , boxed_kernel_func_(nullptr)
+  , unboxed_kernel_func_(nullptr)
+  {}
+
+  bool isValid() const {
+    // TODO We want to introduce the invariant that all kernels must be callable in a boxed way, then this should only check boxed_kernel_func_.
+    return boxed_kernel_func_ != nullptr || unboxed_kernel_func_ != nullptr;
+  }
+
+  /**
+   * Call the function in a boxed way.
+   * If the kernel function was created with an unboxed function,
+   * this will call an unboxing wrapper which then calls into that
+   * unboxed function.
+   *
+   * Example:
+   *
+   * > void boxed_func(OperatorKernel*, Stack* stack) {...}
+   * > KernelFunction func = KernelFunction::makeFromBoxedFunction(&boxed_func);
+   * > Tensor result = func.callBoxed(stack);
+   *
+   * Or, with an unboxed implementation:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   * > Tensor result = func.callBoxed(stack);
+   */
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
```

Comment:
As an organizational point, it would be nice for this error handling code to live in the cpp file rather than the header. There's no reason for it to live inline, and it would make certain types of refactoring/debugging work easier (as it would avoid a recompilation of the world)

Reply:
Calling operators is a hot path that I want to be fully inlineable. I could move it to a KernelFunction_inl.h header if you think that's better? Let me know.

Reply:
I'm talking about the error reporting code, not the calling code.